File: Domain/GameLogic/Rules/PreferHigherDieRule.cs

namespace Domain.GameLogic.Rules
{
    public class PreferHigherDieRule : IMoveSequenceRule
    {
        public IEnumerable<MoveSequence> Apply(
            IEnumerable<MoveSequence> sequences,
            DiceRoll roll)
        {
            var list = sequences.ToList();

            if (list.Count == 0)
            {
                return list;
            }

            var highestDie = list
                .SelectMany(sequence => sequence.Moves)
                .Max(move => move.Die);

            return list
                .Where(s =>
                    s.Moves.Any(move => move.Die == highestDie));
        }
    }
}

File: FRONTEND/backgammon-frontend/src/app/features/game-session/engine/rules/prefer-higher-die.rule.ts
import { MoveSequence } from '../../models/turn/move-sequence.model';
import { MoveSequenceRule } from './move-sequence-rule.model';

export class PreferHigherDieRule implements MoveSequenceRule {
  apply(sequences: MoveSequence[], dice: number[]): MoveSequence[] {
    if (sequences.length === 0) {
      return sequences;
    }

    const maxLength = Math.max(...sequences.map((s) => s.moves.length));

    if (maxLength !== 1) {
      return sequences;
    }

    const highestDice = Math.max(...dice);

    return sequences.filter((seq) =>
      seq.moves.some((m) => m.die === highestDice),
    );
  }
}
